scripts/parse-vcf/parse_vcf.py

```diff
@@ -0,0 +1,408 @@
+    return
+
+
+def chunk_region_config(
+    region_config: RegionConfig, chunk_size
+) -> Iterator[RegionConfig]:
+    """break region config into chunks"""
+    chr = region_config.chr
+    chunks = []
+    for n in range(region_config.start, region_config.end, chunk_size):
+        chunks.append(
+            RegionConfig(
+                chr, n + 1, n + chunk_size, region_config.assembly, region_config.source
+            )
+        )
+    return chunks
+
+
+def fetch_vcf_chunk(config: RegionConfig, path: str) -> List[InsertableAnnotation]:
+    """query a range of a vcf, convert to DataFrame and parse"""
+    print(f"fetching {config}")
+
+    """ note that region may require "chr" prefix to first argument, depending on presence in the contig header """
+    query_result = subprocess.run(
+        [
+            "bcftools",
+            "query",
+            path,
+            "-H",
+            "-r",
+            "{0}:{1}-{2}".format(*config),
+            "--exclude",
+            "INFO/AF>0.02",
+            "-f",
+            "%CHROM\t%POS\t%REF\t%ALT\t%INFO/vep\t%INFO/nhomalt\t%INFO/AC\t%INFO/AF\n",
+        ],
+        capture_output=True,
+        timeout=90,  # prevent process from hanging in case of connectivity issue
+    )
+
+    if query_result.stderr:
+        raise Exception(f"bcftools error!: {query_result.stderr}")
+
+    return query_result.stdout
+
+
+def parse_vcf_response(vcf: bytes):
+    callset_df = cast(
+        pd.DataFrame,
+        pd.read_csv(
+            BytesIO(vcf),
+            sep="\t",
+            na_values=".",
+            dtype={
+                "# [1]CHROM": str,
+                "[2]POS": int,
+                "[3]REF": str,
+                "[4]ALT": str,
+                "[5]vep": str,
+                "[6]nhomalt": int,
+                "[7]AC": int,
+                "[8]AF": float,
+            },
+        ),
+    )
+    """ remove bcftools' brackets from headers and convert to lower case """
+    callset_df.rename(
+        columns=lambda header: re.sub(r"(\[.+\])|\# | \W", "", header).lower(),
+        inplace=True,
+    )
+    callset_df = parse_vep(callset_df)
+
+    callset_df = callset_df.drop_duplicates()
+
+    callset_df["af"] = callset_df["af"].fillna(0)
+
+    return callset_df
+
+
+def parse_vep(df: pd.DataFrame):
+    # FYI
+    VEP_HEADERS = "Allele,Consequence,IMPACT,SYMBOL,Gene,Feature_type,Feature,BIOTYPE,EXON,INTRON,HGVSc,HGVSp,cDNA_position,CDS_position,Protein_position,Amino_acids,Codons,Existing_variation,ALLELE_NUM,DISTANCE,STRAND,FLAGS,VARIANT_CLASS,MINIMISED,SYMBOL_SOURCE,HGNC_ID,CANONICAL,TSL,APPRIS,CCDS,ENSP,SWISSPROT,TREMBL,UNIPARC,GENE_PHENO,SIFT,PolyPhen,DOMAINS,HGVS_OFFSET,GMAF,AFR_MAF,AMR_MAF,EAS_MAF,EUR_MAF,SAS_MAF,AA_MAF,EA_MAF,ExAC_MAF,ExAC_Adj_MAF,ExAC_AFR_MAF,ExAC_AMR_MAF,ExAC_EAS_MAF,ExAC_FIN_MAF,ExAC_NFE_MAF,ExAC_OTH_MAF,ExAC_SAS_MAF,CLIN_SIG,SOMATIC,PHENO,PUBMED,MOTIF_NAME,MOTIF_POS,HIGH_INF_POS,MOTIF_SCORE_CHANGE,LoF,LoF_filter,LoF_flags,LoF_info\n"
```

Comment:
Where do these headers come from?

Reply:
This is a list of the VEP headers that I copy/pasted from (somewhere) for reference.